File: Cpp/374_Guess_Number_Higher_or_Lower.cpp
// Forward declaration of guess API.
// @param num, your guess
// @return -1 if my number is lower, 1 if my number is higher, otherwise return 0
int guess(int num);

class Solution {
public:
    int guessNumber(int n) {
        unsigned int high = n, low = 1, mid = 1;
        while(low<=high){
            mid = (high+low)/2;
            int result = guess(mid);
            if(result == 0) return mid;
            else if(result > 0) low = mid+1;
            else high = mid-1;
        }
        
        return mid;
    }
};

File: Kotlin/374_Guess_Number_Higher_or_Lower.kt
/** 
 * The API guess is defined in the parent class.
 * @param  num   your guess
 * @return 	     -1 if num is lower than the guess number
 *			      1 if num is higher than the guess number
 *               otherwise return 0
 * fun guess(num:Int):Int {}
 */

class Solution:GuessGame() {
    override fun guessNumber(n:Int):Int {
        var lo = 1
        var hi = n
        
        while (lo <= hi) {
            val mid = (hi - lo) / 2 + lo
            val result = guess(mid)
            
            when(result) {
                0 -> return mid
                1 -> lo = mid + 1
                -1 -> hi = mid - 1
            }
        }
        
        return lo
    }
}